csrc/device_lower/analysis/tensor_memory.cpp

```diff
@@ -183,6 +188,178 @@ TMemAlllocationInfo computeTMemAlllocationInfo(Fusion* fusion) {
   return result;
 }
 
+// Get the TID Parallel types that are not trivial. We are not interested
+// in the parallel types that are not used in the kernel, and the ones that have
+// size 1. The order of the returned parallel types is from z to x.
+std::vector<ParallelType> getNonTrivialActiveThreadParallelTypes(
+    Fusion* fusion) {
+  const auto& pdim_map = GpuLower::current()->parallelDimensionMap();
+  std::vector<ParallelType> nontrivial_tid_ptypes;
+  for (auto pt : std::views::reverse(kParallelTypeTIDs)) {
+    Val* size = pdim_map.getRaw(pt);
+    if (size == nullptr) {
+      continue;
+    }
+    Val* size_is_one =
+        simplifyExpr(SimplifyingIrBuilder::eqExpr(size, fusion->oneVal()));
+    if (size_is_one->isTrue()) {
+      continue;
+    }
+    nontrivial_tid_ptypes.push_back(pt);
+  }
+
+  NVF_CHECK(
+      !nontrivial_tid_ptypes.empty(),
+      "Invalid data access pattern in TMem load/store: ",
+      "TMem load/store must be warp-collective, but CTA size is one.");
+  return nontrivial_tid_ptypes;
+}
+
+// Get the [TIDz, TIDy, TIDx] projected to the given expression as ValGroups,
+// and merge them by contiguity. If any of the TIDz, TIDy, TIDx is not
+// interested (see above), we just ignore it. Return the merged ValGroups as an
+// AbstractTensor.
+//
+// Why do we need this function?
+//
+// In the CUDA programming model, each CTA has TIDx, TIDy, and TIDz.
+// Unfortunately, the mapping of these TIDs to hardware concepts like warp, warp
+// group, are not clear and depend on the kernel launch configuration. Here, we
+// try to not assume anything like "TIDx must be a multiple of 32", but still,
+// we must be able to validate and pattern match the data access of the tensor
+// memory load/store.
+//
+// We need to construct a ValGroup that represents "warp" for an expression from
+// its consumer's loop domain. Naively speaking, it is just:
+//   split(Iz * Iy * Ix, 32).inner
```

Comment:
Are we assuming these splits are always divisible?

Reply:
Yes, and we are validating this below https://github.com/NVIDIA/Fuser/pull/4015/files#diff-e01895ae709d394f4be39fdae0a376c766bc2cecbb1e5287d568eee37ac92bc1R420:
```C++
    GpuLower::current()->validate(
        inner_extent_is_multiple_of_32,
        "Invalid data access pattern in TMem load/store: ",
        "TMem load/store must be warp-collective, but the innermost extent is not a multiple of 32.");
```